File: go/0683_k_empty_slots.go
func kEmptySlots(flowers []int, k int) int {
	bit, isBloom := make(BIT, len(flowers)+1), make([]int, len(flowers))
	for i, f := range flowers {
		f-- // adjut to (0..N)
		add(bit, f, 1)
		isBloom[f] = 1
		if (f-k-1 >= 0 && isBloom[f-k-1] == 1 && sum(bit, f-1)-sum(bit, f-k-1) == 0) ||
			(f+k+1 < len(flowers) && isBloom[f+k+1] == 1 && sum(bit, f+k)-sum(bit, f) == 0) {
			return i + 1
		}
	}
	return -1
}

type BIT []int

func add(bit BIT, index, value int) {
	for index = index + 1; index < len(bit); {
		bit[index] += value
		index += index & -index
	}
}

func sum(bit BIT, index int) (sum int) {
	for index = index + 1; index > 0; {
		sum += bit[index]
		index -= index & -index
	}
	return sum
}

File: python/0683_k_empty_slots.py
class Solution:
    def kEmptySlots(self, flowers, k):
        """
        :type flowers: List[int]
        :type k: int
        :rtype: int
        """

        days = [0] * len(flowers)
        for day, flower in enumerate(flowers):
            days[flower - 1] = day

        res = -1
        left, right = 0, k + 1
        while right < len(days):
            later = max(days[left], days[right])
            satisfied = True

            for i in range(left + 1, right):
                if days[i] <= later:
                    left, right = i, i + k + 1
                    satisfied = False
                    break

            if not satisfied:
                continue

            res = min(later + 1, res) if res != -1 else later + 1
            left, right = right, right + k + 1
        return res